src/interface/swarm_pack_base.hpp

```diff
@@ -0,0 +1,267 @@
+    pack.bounds_ = bounds_t("bounds", bounds_leading_size, nblocks, nvar);
+    auto bounds_h = Kokkos::create_mirror_view(pack.bounds_);
+
+    // Fill the views
+    ForEachBlock(pmd, std::vector<bool>{}, [&](int b, auto *pmbd) {
+      int idx = 0;
+      auto swarm = pmbd->GetSwarm(desc.swarm_name);
+      for (int i = 0; i < nvar; ++i) {
+        bounds_h(0, b, i) = idx;
+        for (auto &pv : swarm->template GetVariableVector<TYPE>()) {
+          if (desc.IncludeVariable(i, pv)) {
+            for (int t = 0; t < pv->GetDim(6); ++t) {
+              for (int u = 0; u < pv->GetDim(5); ++u) {
+                for (int v = 0; v < pv->GetDim(4); ++v) {
+                  for (int l = 0; l < pv->GetDim(3); ++l) {
+                    for (int m = 0; m < pv->GetDim(2); ++m) {
+                      pack_h(0, b, idx) = pv->data.Get(0, t, u, v, l, m);
+                      PARTHENON_REQUIRE(
+                          pack_h(0, b, idx).size() > 0,
+                          "Seems like this variable might not actually be allocated.");
+                      idx++;
+                    }
+                  }
+                }
+              }
+            }
+            bounds_h(2, b, i) = (pv->GetDim(6) > 1 ? 1 : 0) +
+                                (pv->GetDim(5) > 1 ? 1 : 0) +
+                                (pv->GetDim(4) > 1 ? 1 : 0) +
+                                (pv->GetDim(3) > 1 ? 1 : 0) + (pv->GetDim(2) > 1 ? 1 : 0);
+            for (int d = 0; d < bounds_h(2, b, i); d++) {
+              bounds_h(3 + d, b, i) = pv->GetDim(2 + d);
+            }
+          }
+        }
+
+        bounds_h(1, b, i) = idx - 1;
+
+        if (bounds_h(1, b, i) < bounds_h(0, b, i)) {
+          // Did not find any allocated variables meeting our criteria
+          bounds_h(0, b, i) = -1;
+          // Make the upper bound more negative so a for loop won't iterate once
+          bounds_h(1, b, i) = -2;
+        }
+      }
+    });
+
+    Kokkos::deep_copy(pack.pack_, pack_h);
+    Kokkos::deep_copy(pack.bounds_, bounds_h);
+
+    pack.contexts_ = contexts_t("contexts", nblocks);
+    pack.contexts_h_ = Kokkos::create_mirror_view(pack.contexts_);
+    pack.max_active_indices_ = max_active_indices_t("max_active_indices", nblocks);
+    pack.flat_index_map_ = max_active_indices_t("flat_index_map", nblocks + 1);
+    BuildSupplemental(pmd, desc, pack);
+
+    return pack;
+  }
+
+  template <class T>
+  static SwarmPackBase<TYPE> BuildAndAdd(T *pmd, const SwarmPackDescriptor<TYPE> &desc) {
+    auto &pack_map = pmd->template GetSwarmPackCache<TYPE>().pack_map;
+    pack_map[desc.identifier] = Build<T>(pmd, desc);
+    return pack_map[desc.identifier];
+  }
+
+  template <class T>
+  static SwarmPackBase<TYPE> Get(T *pmd, const impl::SwarmPackDescriptor<TYPE> &desc) {
+    auto &pack_map = pmd->template GetSwarmPackCache<TYPE>().pack_map;
+    if (pack_map.count(desc.identifier) > 0) {
```

Comment:
I guess swarm packs are static aside from the supplemental part, so if they exist in the cache they must be fresh?

Reply:
correct